File: generated_code/python/test_unittest_908.py
import os
import sys
import importlib.util
import unittest

# Dynamic loading of the solution module as per guidelines
solution_path = os.environ.get("TEST_SOLUTION_FILE")
if not solution_path:
    raise RuntimeError("TEST_SOLUTION_FILE environment variable is not set.")

spec = importlib.util.spec_from_file_location("Solution", solution_path)
sol_module = importlib.util.module_from_spec(spec)
spec.loader.exec_module(sol_module)
Solution = sol_module.Solution

# Dynamically access ListNode from the module if defined there, otherwise define a standard one
if hasattr(sol_module, "ListNode"):
    ListNode = sol_module.ListNode
else:
    class ListNode:
        def __init__(self, val=0, next=None):
            self.val = val
            self.next = next


class TestMiddleNode(unittest.TestCase):
    def _create_linked_list(self, arr):
        """Helper method to construct a linked list from an array."""
        if not arr:
            return None
        head = ListNode(arr[0])
        current = head
        for val in arr[1:]:
            current.next = ListNode(val)
            current = current.next
        return head

    def _linked_list_to_list(self, head):
        """Helper method to convert a linked list back to an array for assertion."""
        result = []
        current = head
        while current:
            result.append(current.val)
            current = current.next
        return result

    def test_example_1_odd_length(self):
        """Test with an odd number of nodes (Example 1: [1,2,3,4,5] -> [3,4,5])."""
        input_list = [1, 2, 3, 4, 5]
        head = self._create_linked_list(input_list)
        
        sol = Solution()
        actual_middle = sol.middleNode(head)
        
        actual_output = self._linked_list_to_list(actual_middle)
        expected_output = [3, 4, 5]
        self.assertEqual(actual_output, expected_output)

    def test_example_2_even_length(self):
        """Test with an even number of nodes (Example 2: [1,2,3,4,5,6] -> [4,5,6])."""
        input_list = [1, 2, 3, 4, 5, 6]
        head = self._create_linked_list(input_list)
        
        sol = Solution()
        actual_middle = sol.middleNode(head)
        
        actual_output = self._linked_list_to_list(actual_middle)
        expected_output = [4, 5, 6]
        self.assertEqual(actual_output, expected_output)

    def test_edge_case_single_node(self):
        """Test the minimum boundary constraint: a list containing exactly 1 node."""
        input_list = [1]
        head = self._create_linked_list(input_list)
        
        sol = Solution()
        actual_middle = sol.middleNode(head)
        
        actual_output = self._linked_list_to_list(actual_middle)
        expected_output = [1]
        self.assertEqual(actual_output, expected_output)

    def test_edge_case_two_nodes(self):
        """Test a minimal even length list containing exactly 2 nodes."""
        input_list = [1, 2]
        head = self._create_linked_list(input_list)
        
        sol = Solution()
        actual_middle = sol.middleNode(head)
        
        actual_output = self._linked_list_to_list(actual_middle)
        expected_output = [2]
        self.assertEqual(actual_output, expected_output)

    def test_edge_case_maximum_constraints(self):
        """Test the maximum boundary constraint: a list containing 100 nodes."""
        input_list = list(range(1, 101))  # [1, 2, ..., 100]
        head = self._create_linked_list(input_list)
        
        sol = Solution()
        actual_middle = sol.middleNode(head)
        
        actual_output = self._linked_list_to_list(actual_middle)
        # Even length 100: middle nodes are 50 and 51, should return from 51 onwards
        expected_output = list(range(51, 101))
        self.assertEqual(actual_output, expected_output)


if __name__ == "__main__":
    unittest.main()

File: generated_code/cpp/test_unittest_908.cpp
#include <gtest/gtest.h>
#include <vector>
#include "solution_proxy.h"

// Helper function to create a linked list from a vector
ListNode* createList(const std::vector<int>& values) {
    if (values.empty()) return nullptr;
    ListNode* head = new ListNode(values[0]);
    ListNode* current = head;
    for (size_t i = 1; i < values.size(); ++i) {
        current->next = new ListNode(values[i]);
        current = current->next;
    }
    return head;
}

// Helper function to free the allocated memory of a linked list
void freeList(ListNode* head) {
    while (head != nullptr) {
        ListNode* temp = head;
        head = head->next;
        delete temp;
    }
}

// Test case for Example 1: Odd number of nodes
TEST(MiddleNodeTest, Example1_OddNumberOfNodes) {
    Solution solution;
    ListNode* head = createList({1, 2, 3, 4, 5});
    
    ListNode* result = solution.middleNode(head);
    
    ASSERT_NE(result, nullptr);
    EXPECT_EQ(result->val, 3);
    
    freeList(head);
}

// Test case for Example 2: Even number of nodes (should return second middle)
TEST(MiddleNodeTest, Example2_EvenNumberOfNodes) {
    Solution solution;
    ListNode* head = createList({1, 2, 3, 4, 5, 6});
    
    ListNode* result = solution.middleNode(head);
    
    ASSERT_NE(result, nullptr);
    EXPECT_EQ(result->val, 4);
    
    freeList(head);
}

// Edge case: Minimum number of nodes (exactly 1 node)
TEST(MiddleNodeTest, EdgeCase_SingleNodeList) {
    Solution solution;
    ListNode* head = createList({1});
    
    ListNode* result = solution.middleNode(head);
    
    ASSERT_NE(result, nullptr);
    EXPECT_EQ(result->val, 1);
    EXPECT_EQ(result->next, nullptr);
    
    freeList(head);
}

// Edge case: Smallest even list (exactly 2 nodes)
TEST(MiddleNodeTest, EdgeCase_TwoNodesList) {
    Solution solution;
    ListNode* head = createList({1, 2});
    
    ListNode* result = solution.middleNode(head);
    
    ASSERT_NE(result, nullptr);
    EXPECT_EQ(result->val, 2);
    EXPECT_EQ(result->next, nullptr);
    
    freeList(head);
}

// Edge case: Maximum scale constraint check (100 nodes, odd-like boundary evaluation via 99 nodes)
TEST(MiddleNodeTest, ConstraintBoundary_99Nodes) {
    Solution solution;
    std::vector<int> values;
    for (int i = 1; i <= 99; ++i) {
        values.push_back(i);
    }
    ListNode* head = createList(values);
    
    ListNode* result = solution.middleNode(head);
    
    ASSERT_NE(result, nullptr);
    EXPECT_EQ(result->val, 50); // Middle of 99 is the 50th node (1-indexed)
    
    freeList(head);
}

// Edge case: Maximum allowed nodes according to constraints (100 nodes)
TEST(MiddleNodeTest, ConstraintBoundary_100Nodes) {
    Solution solution;
    std::vector<int> values;
    for (int i = 1; i <= 100; ++i) {
        values.push_back(i);
    }
    ListNode* head = createList(values);
    
    ListNode* result = solution.middleNode(head);
    
    ASSERT_NE(result, nullptr);
    EXPECT_EQ(result->val, 51); // Second middle of 100 is the 51st node (1-indexed)
    
    freeList(head);
}